From 300e7cfa6e924f30d0d5721cf68fcd78423dd585 Mon Sep 17 00:00:00 2001
From: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
Date: Wed, 31 Jul 2013 12:32:48 +0900
Subject: [PATCH 0811/1083] ASoC: sh: scu: Add variable period size support

The period size and the number of periods are changed.

Signed-off-by: Shunsuke Kataoka <shunsuke.kataoka.df@renesas.com>
---
 include/sound/sh_scu.h | 10 +++++-----
 sound/soc/sh/scu_pcm.c |  6 +++---
 2 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/include/sound/sh_scu.h b/include/sound/sh_scu.h
index 03821b6..b1da2f0 100644
--- a/include/sound/sh_scu.h
+++ b/include/sound/sh_scu.h
@@ -33,11 +33,11 @@
  *	ALSA
  */
 /* buffer information */
-#define SCU_PERIOD_BYTES_MIN	(16 * 1024)
-#define SCU_PERIOD_BYTES_MAX	(16 * 1024)
-#define SCU_PERIODS_MIN		4
-#define SCU_PERIODS_MAX		4
-#define SCU_BUFFER_BYTES_MAX	(SCU_PERIOD_BYTES_MAX * SCU_PERIODS_MAX)
+#define SCU_BUFFER_BYTES_MAX	(32 * 1024)
+#define SCU_PERIOD_BYTES_MIN	512
+#define SCU_PERIOD_BYTES_MAX	8192
+#define SCU_PERIODS_MIN		(SCU_BUFFER_BYTES_MAX / SCU_PERIOD_BYTES_MAX)
+#define SCU_PERIODS_MAX		(SCU_BUFFER_BYTES_MAX / SCU_PERIOD_BYTES_MIN)
 
 /* scu dapm route information */
 /* playback route */
diff --git a/sound/soc/sh/scu_pcm.c b/sound/soc/sh/scu_pcm.c
index 541717f..4cc4345 100644
--- a/sound/soc/sh/scu_pcm.c
+++ b/sound/soc/sh/scu_pcm.c
@@ -76,7 +76,7 @@ static void scu_dma_callback(struct snd_pcm_substream *ss)
 
 	FNC_ENTRY
 
-	buf_pos = pcminfo->period & (SCU_PERIODS_MAX - 1);
+	buf_pos = pcminfo->period % runtime->periods;
 	dma_size = frames_to_bytes(runtime, runtime->period_size);
 	dma_paddr = runtime->dma_addr + (buf_pos * dma_size);
 	dma_sync_single_for_cpu(dai->dev, dma_paddr, dma_size, DMA_DIR(dir));
@@ -308,7 +308,7 @@ static int scu_audma_start(int sid, struct snd_pcm_substream *ss)
 	dai = scu_get_dai(ss);
 
 	/* buffer control */
-	buf_pos = pcminfo->period & (SCU_PERIODS_MAX - 1);
+	buf_pos = pcminfo->period % runtime->periods;
 	DBG_MSG("buf_pos=%d\n", buf_pos);
 
 	/* DMA size */
@@ -742,7 +742,7 @@ static snd_pcm_uframes_t scu_pcm_pointer_dma(struct snd_pcm_substream *ss)
 	snd_pcm_uframes_t position = 0;
 
 	position = runtime->period_size *
-			(pcminfo->tran_period & (SCU_PERIODS_MAX - 1));
+			(pcminfo->tran_period % runtime->periods);
 
 	DBG_MSG("\tposition = %d\n", (u32)position);
 
-- 
1.8.3.2

